ophyd/status.py

```diff
@@ -127,25 +184,118 @@ def _settled(self):
+            # And it's possible that in this space be between the above
+            # statement timing out grabbing the lock just below,
+            # set_exception(exc) has been called. Both of these possibilties
+            # are accounted for.
+            logger.warning("%r has timed out", self)
+            with self._externally_initiated_completion_lock:
+                # Set the exception and mark the Status as done, unless
+                # set_exception(exc) was called externally before we grabbed
+                # the lock.
+                if self._exception is None:
+                    exc = StatusTimeoutError(
+                        f"Status {self!r} failed to complete in specified timeout.")
+                    self._exception = exc
+        # Now we know whether or not we have succeed or failed, either by
+        # timeout above or by set_exception(exc), so we can set the Event that
+        # will mark this Status as done.
         with self._lock:
-            if self.done:
-                # We timed out while waiting for the settle time.
+            self._event.set()
+        if self._exception is not None:
+            try:
+                self._handle_failure()
+            except Exception:
+                logger.exception(
+                    "Failure handling on %r raised an error.", self)
+        # Rain or shine, mark this as "settled" and run the callbacks.
+        # The callbacks have access to self, from which they can distinguish
+        # success or failure.
+        self._settled()
+        for cb in self._callbacks:
+            cb(self)
+        self._callbacks.clear()
+
+    def set_exception(self, exc):
+        """
+        Mark as finished but failed with the given Exception.
+
+        This method should generally not be called by the *recipient* of this
+        Status object, but only by the object that created and returned it.
+
+        Parameters
+        ----------
+        exc: Exception
+        """
+        # Since we rely on this being raise-able later, check proactively to
+        # avoid potentially very confusing failures.
+        if not (isinstance(exc, Exception)
+                or isinstance(exc, type) and issubclass(exc, Exception)):
+            # Note that Python allows `raise Exception` or raise Exception()`
+            # so we allow a class or an instance here too.
+            raise ValueError(f"Expected an Exception, got {exc!r}")
+
+        # Ban certain Timeout subclasses that have special significance. This
+        # would probably never come up except due to some rare user error, but
+        # if it did it could be very confusing indeed!
+        for exc_class in (StatusTimeoutError, WaitTimeoutError):
+            if (isinstance(exc, exc_class)
+                    or isinstance(exc, type) and issubclass(exc, exc_class)):
+                raise ValueError(
+                    f"{exc_class} as special significance and cannot be set "
```

Comment:
```suggestion
                    f"{exc_class} has special significance and cannot be set "
```

Reply:
This suggestion is rendering incorrectly in my browser, so if it also is rendering incorrectly for you I was trying to suggest:
as -> has